File: app/lib/core/models/dungeon_map.dart
import 'room.dart';

class DungeonMap {
  const DungeonMap({
    required this.seed,
    required this.width,
    required this.height,
    required this.rooms,
    required this.entranceRoomId,
  });

  factory DungeonMap.fromJson(Map<String, dynamic> json) => DungeonMap(
        seed: json['seed'] as String,
        width: json['width'] as int,
        height: json['height'] as int,
        entranceRoomId: json['entranceRoomId'] as String,
        rooms: (json['rooms'] as Map).map(
          (key, value) => MapEntry(
            key as String,
            Room.fromJson(value as Map<String, dynamic>),
          ),
        ),
      );

  final String seed;

  /// Largura da grade do autômato celular, em número de células.
  final int width;

  /// Altura da grade do autômato celular, em número de células.
  final int height;

  /// Todas as células da grade (width × height), indexadas por "{x},{y}",
  /// inclusive as do tipo [RoomType.wall].
  final Map<String, Room> rooms;
  final String entranceRoomId;

  /// Busca a sala na posição (x, y), ou `null` se estiver fora da grade.
  Room? roomAt(int x, int y) => rooms['$x,$y'];
}

File: backend/src/main/java/com/furios/labirinto_dungeonico/dungeon/DungeonMap.java
package com.furios.labirinto_dungeonico.dungeon;

import com.fasterxml.jackson.annotation.JsonAutoDetect;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * fieldVisibility ANY: expõe os campos diretamente ao Jackson, já que a classe usa acessores
 * fluentes (seed(), rooms()) em vez do padrão getX() que o Jackson reconhece por padrão.
 */
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
public class DungeonMap {

    private final String seed;
    private final String entranceRoomId;
    private final int width;
    private final int height;
    private final Map<String, Room> rooms = new LinkedHashMap<>();

    public DungeonMap(String seed, String entranceRoomId, int width, int height) {
        this.seed = seed;
        this.entranceRoomId = entranceRoomId;
        this.width = width;
        this.height = height;
    }

    public String seed() {
        return seed;
    }

    public String entranceRoomId() {
        return entranceRoomId;
    }

    /** Largura da grade do autômato celular, em número de células. */
    public int width() {
        return width;
    }

    /** Altura da grade do autômato celular, em número de células. */
    public int height() {
        return height;
    }

    /**
     * Todas as células da grade (width × height), indexadas por "{x},{y}", inclusive as do
     * tipo WALL. Um app cliente pode reconstruir a grade completa iterando x de 0..width-1 e
     * y de 0..height-1 e buscando "{x},{y}" neste mapa.
     */
    public Map<String, Room> rooms() {
        return rooms;
    }
}
